# Task 2: Multivariable Regression

**The Goal**: Introduce complexity and multi-dimensional data.

**The Action**: Upgrade the model to predict Temperature using multiple continuous variables simultaneously. Based on our dataset (`ai4i2020.csv`), we will use `Rotational speed [rpm]`, `Torque [Nm]`, `Tool wear [min]`, and `Air temperature [K]`.

**The Research**: Compare the MSE of this new model against the simple model from Task 1 (MSE ~ 2.1691), and document why adding more relevant features usually reduces the error rate.

In [1]:
import pandas as pd
from sklearn.model_selection import train_test_split
from sklearn.linear_model import LinearRegression
from sklearn.metrics import mean_squared_error

# Load dataset
df = pd.read_csv('ai4i2020.csv')

# Select multiple continuous features as predictors
features = ['Rotational speed [rpm]', 'Torque [Nm]', 'Tool wear [min]', 'Air temperature [K]']
X = df[features]
y = df['Process temperature [K]']

# Split data into train and test sets
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)

# Train the Multivariable Regression model
model = LinearRegression()
model.fit(X_train, y_train)

# Make predictions
y_pred = model.predict(X_test)

# Calculate Mean Squared Error (MSE)
mse_multi = mean_squared_error(y_test, y_pred)
mse_simple = 2.1691 # from Task 1

print(f'Task 1 Simple Linear Regression MSE: {mse_simple:.4f}')
print(f'Task 2 Multivariable Regression MSE: {mse_multi:.4f}')
print(f'Improvement in MSE: {mse_simple - mse_multi:.4f}')

Task 1 Simple Linear Regression MSE: 2.1691
Task 2 Multivariable Regression MSE: 0.5299
Improvement in MSE: 1.6392


### Research & Documentation

**Comparison of Results:**
*   **Simple Linear Regression MSE (Task 1):** ~2.1691
*   **Multivariable Regression MSE (Task 2):** ~0.5299

The error profile dropped significantly (by ~1.6392) when transitioning from a single variable to a multi-variable architecture.

**Why adding more relevant features usually reduces the error rate:**
When predicting a complex physical outcome—like internal machine temperature—relying on a single variable (like solely `RPM`) captures only one dimension of the operational physics. Real-world target variables are almost always influenced simultaneously by several drivers.

By feeding the model additional, relevant dimensions (such as `Torque`, `Tool wear`, and environmental `Air temperature`), we equip our model to observe the larger, complete system state. It calculates independent coefficients (weights) for each physical aspect, enabling the model to account for variances that would otherwise remain "unexplained noise" in a simpler scenario. Consequently, because the model's understanding of the machine is more fully contextually aware, its predictive accuracy increases, therefore lowering the MSE.

*(Note: Adding features unconditionally isn't always ideal due to the risk of "overfitting" or "multicollinearity", but adding highly relevant distinct signal features generally reduces prediction loss.)*